system/src/commonMain/kotlin/com/splendo/kaluga/system/network/state/NetworkStateRepo.kt

```diff
@@ -0,0 +1,123 @@
+    You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+ */
+
+package com.splendo.kaluga.system.network.state
+
+import co.touchlab.stately.concurrency.AtomicReference
+import com.splendo.kaluga.state.ColdStateRepo
+import com.splendo.kaluga.system.network.BaseNetworkManager
+import com.splendo.kaluga.system.network.Network
+import com.splendo.kaluga.system.network.unknownNetworkOf
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.launch
+
+class NetworkStateRepo(
+    private val networkManagerBuilder: BaseNetworkManager.Builder,
+) : ColdStateRepo<NetworkState>() {
+
+    interface Builder {
+        fun create(): NetworkStateRepo
+    }
+
+    private var _lastKnownNetwork = AtomicReference<Network>(Network.Unknown.WithoutLastNetwork(Network.Unknown.Reason.NOT_CLEAR))
+    internal var lastKnownNetwork: Network
+        get() = _lastKnownNetwork.get()
+        set(value) { _lastKnownNetwork.set(value) }
+
+    internal var networkManager: BaseNetworkManager? = null
+        set(value) {
+            field?.let {
+                if(value == null) {
+                    it.dispose()
+                }
+            }
+            field = value
+        }
+
+    override suspend fun deinitialize(state: NetworkState) {
+        lastKnownNetwork = state.networkType
+        networkManager = null
+    }
+
+    override suspend fun initialValue(): NetworkState {
+        networkManager = networkManagerBuilder.create(::onNetworkStateChange)
+        return when(lastKnownNetwork) {
+            is Network.Unknown.WithoutLastNetwork -> {
+                with (lastKnownNetwork as Network.Unknown.WithoutLastNetwork) {
+                    NetworkState.Unknown(unknownNetworkOf(reason), networkManager!!)
+                }
+            }
+            is Network.Unknown.WithLastNetwork -> {
+                with (lastKnownNetwork as Network.Unknown.WithLastNetwork) {
+                    NetworkState.Unknown(unknownNetworkOf(reason), networkManager!!)
+                }
+            }
+            is Network.Known.Cellular -> NetworkState.Available(Network.Known.Cellular(), networkManager!!)
+            is Network.Known.Wifi -> {
+                with(lastKnownNetwork as Network.Known.Wifi) {
+                    NetworkState.Available(Network.Known.Wifi(isExpensive), networkManager!!)
+                }
+            }
+            Network.Known.Absent -> NetworkState.Unavailable(Network.Known.Absent, networkManager!!)
+        }
+    }
+
+    internal fun onNetworkStateChange(network: Network) {
+        launch {
```

Comment:
Should I pass a `Job`?

Reply:
There is two options, just use `runBlocking` with the correct context (that of the repo probably) and call the suspended function. This will block this thread, which is probably a main thread, but it should still be fast since I think nothing will suspend for a long time (but it's your code, so you) make sure.

If your idea was to NOT block on this thread and that's why you use launch (which, well, will run on the main thread anyway, just later) then you need to use a buffered Channel (probably best to through a `Flow`) to preserve the order of these events coming in.

Reply:
1650407